File: scheduling/employee_scheduler.js
// Employee Scheduler 
// Deterministic RNG (mulberry32)
function mulberry32(a) {
  return function() {
    var t = a += 0x6D2B79F5;
    t = Math.imul(t ^ t >>> 15, t | 1);
    t ^= t + Math.imul(t ^ t >>> 7, t | 61);
    return ((t ^ t >>> 14) >>> 0) / 4294967296;
  }
}
const rand = mulberry32(42);
function shuffle(arr){ for(let i=arr.length-1;i>0;i--){ const j=Math.floor(rand()*(i+1)); [arr[i],arr[j]]=[arr[j],arr[i]]; } }

const DAYS = ["Mon","Tue","Wed","Thu","Fri","Sat","Sun"];
const SHIFTS = ["Morning","Afternoon","Evening"];
const SHIFT_MIN_STAFF = 2;
const SHIFT_CAPACITY = 2; // set to 2 if you want the simplest guaranteed pass
const MAX_DAYS_PER_EMP = 5;

const EMPLOYEE_PREFERENCES = {
  "Ava":   ["Morning","Afternoon","Evening"],
  "Ben":   ["Afternoon","Evening","Morning"],
  "Cara":  ["Evening","Morning","Afternoon"],
  "Diego": ["Morning","Evening","Afternoon"],
  "Eli":   ["Afternoon","Morning","Evening"],
  "Farah": ["Evening","Afternoon","Morning"],
  "Grace": ["Morning","Afternoon","Evening"],
  "Hassan":["Afternoon","Evening","Morning"],
  "Ivy":   ["Morning","Evening","Afternoon"],
  "Jon":   ["Evening","Morning","Afternoon"],
};

function emptySchedule(){
  const sch = {};
  for(const d of DAYS){
    sch[d] = {};
    for(const s of SHIFTS) sch[d][s] = [];
  }
  return sch;
}
function employeeStats(names){
  const st = {};
  for(const n of names) st[n] = {days:0, assigned:new Set()};
  return st;
}
function canWork(emp, day, stats){
  return stats[emp].days < MAX_DAYS_PER_EMP && !stats[emp].assigned.has(day);
}
function assign(emp, day, shift, schedule, stats){
  schedule[day][shift].push(emp);
  stats[emp].days += 1;
  stats[emp].assigned.add(day);
}
function hasCapacity(day, shift, schedule){
  return schedule[day][shift].length < SHIFT_CAPACITY;
}
function anyCapacity(day, schedule){
  return SHIFTS.some(s => schedule[day][s].length < SHIFT_CAPACITY);
}

function buildSchedule(){
  const schedule = emptySchedule();
  const employees = Object.keys(EMPLOYEE_PREFERENCES);
  const stats = employeeStats(employees);
  let carryOver = [];

  for(const day of DAYS){
    const pool = carryOver.concat(employees);
    carryOver = [];
    shuffle(pool);

    // First pass: try preferences with capacity
    for(const emp of pool){
      if(!canWork(emp, day, stats)) continue;
      let placed = false;
      for(const pref of EMPLOYEE_PREFERENCES[emp]){
        if(hasCapacity(day, pref, schedule)){
          assign(emp, day, pref, schedule, stats);
          placed = true; break;
        }
      }
      if(!placed){
        for(const alt of SHIFTS){
          if(hasCapacity(day, alt, schedule)){
            assign(emp, day, alt, schedule, stats);
            placed = true; break;
          }
        }
      }
      if(!placed) carryOver.push(emp);
    }

    // Minimum staffing with fairness (fewest days first)
    for(const shift of SHIFTS){
      let need = Math.max(0, SHIFT_MIN_STAFF - schedule[day][shift].length);
      if(need === 0) continue;
      let eligible = employees.filter(e => canWork(e, day, stats));
      eligible.sort((a,b) => stats[a].days - stats[b].days);
      while(need-- > 0 && eligible.length){
        const pick = eligible.shift();
        assign(pick, day, shift, schedule, stats);
        // keep list sorted as days change
        eligible.sort((a,b) => stats[a].days - stats[b].days);
      }
    }

    // Optional second pass to fill remaining capacity by fairness
    let unassigned = employees.filter(e => canWork(e, day, stats));
    unassigned.sort((a,b) => stats[a].days - stats[b].days);
    for(const emp of unassigned){
      let placed = false;
      for(const pref of EMPLOYEE_PREFERENCES[emp]){
        if(hasCapacity(day, pref, schedule)){
          assign(emp, day, pref, schedule, stats);
          placed = true; break;
        }
      }
      if(!placed && !anyCapacity(day, schedule)) carryOver.push(emp);
    }
  }
  return {schedule, stats};
}

function printSchedule(schedule, stats){
  console.log("\n=== FINAL WEEKLY SCHEDULE ===");
  const header = `Day  | ${SHIFTS.map(s => s.padEnd(25)).join(" | ")}`;
  console.log(header);
  console.log("-".repeat(header.length));
  for(const day of DAYS){
    const row = SHIFTS.map(s => {
      const v = schedule[day][s];
      return (v.length ? v.join(", ") : "-").padEnd(25);
    }).join(" | ");
    console.log(`${day.padEnd(4)} | ${row}`);
  }

  console.log("\n=== EMPLOYEE SUMMARY (max 5 days) ===");
  const names = Object.keys(stats).sort();
  for(const n of names){
    const days = stats[n].days;
    const ds = Array.from(stats[n].assigned);
    ds.sort((a,b) => DAYS.indexOf(a) - DAYS.indexOf(b));
    console.log(`${n.padEnd(8)}: ${days} day(s) \u2192 ${ds.join(", ")}`);
  }

  // Checks
  console.log("\n=== CONSTRAINT CHECKS ===");
  // One shift per day
  let okOne = true;
  for(const day of DAYS){
    const seen = new Set();
    for(const s of SHIFTS){
      for(const e of schedule[day][s]){
        if(seen.has(e)) okOne = false;
        seen.add(e);
      }
    }
  }
  console.log("One shift per employee per day:", okOne ? "OK" : "VIOLATION");

  // Max 5 per week
  const okMax = Object.values(stats).every(v => v.days <= MAX_DAYS_PER_EMP);
  console.log("Max 5 days per employee:", okMax ? "OK" : "VIOLATION");

  // Minimum 2 per shift
  let minOk = true;
  for(const d of DAYS){
    for(const s of SHIFTS){
      if(schedule[d][s].length < SHIFT_MIN_STAFF) minOk = false;
    }
  }
  console.log("Minimum 2 employees per shift:", minOk ? "OK" : "COULD NOT FILL");
}

// Run
const {schedule, stats} = buildSchedule();
printSchedule(schedule, stats);

File: scheduling/employee_scheduler.py
# Employee Scheduler  
import random
random.seed(42)

DAYS = ["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"]
SHIFTS = ["Morning", "Afternoon", "Evening"]

SHIFT_MIN_STAFF = 2          # Requirement: at least 2 employees per shift
SHIFT_CAPACITY  = 2          # Soft capacity per shift (set to 2 if you prefer)
MAX_DAYS_PER_EMP = 5         # At most 5 days per employee per week

# ---------- Sample data ----------
# Preferences are ranked lists (1st → last) and reused for each day
EMPLOYEE_PREFERENCES = {
    "Ava":    ["Morning", "Afternoon", "Evening"],
    "Ben":    ["Afternoon", "Evening", "Morning"],
    "Cara":   ["Evening", "Morning", "Afternoon"],
    "Diego":  ["Morning", "Evening", "Afternoon"],
    "Eli":    ["Afternoon", "Morning", "Evening"],
    "Farah":  ["Evening", "Afternoon", "Morning"],
    "Grace":  ["Morning", "Afternoon", "Evening"],
    "Hassan": ["Afternoon", "Evening", "Morning"],
    "Ivy":    ["Morning", "Evening", "Afternoon"],
    "Jon":    ["Evening", "Morning", "Afternoon"],
}

# ---------- Helpers ----------
def empty_schedule():
    return {day: {shift: [] for shift in SHIFTS} for day in DAYS}

def employee_stats(names):
    return {name: {"days": 0, "assigned_days": set()} for name in names}

def can_work(emp, day, stats):
    return (stats[emp]["days"] < MAX_DAYS_PER_EMP) and (day not in stats[emp]["assigned_days"])

def assign(emp, day, shift, schedule, stats):
    schedule[day][shift].append(emp)
    stats[emp]["days"] += 1
    stats[emp]["assigned_days"].add(day)

def has_capacity(day, shift, schedule):
    return len(schedule[day][shift]) < SHIFT_CAPACITY

def any_capacity(day, schedule):
    return any(len(schedule[day][s]) < SHIFT_CAPACITY for s in SHIFTS)

# ---------- Build schedule ----------
def build_schedule():
    schedule = empty_schedule()
    employees = list(EMPLOYEE_PREFERENCES.keys())
    stats = employee_stats(employees)

    carry_over = []  # employees bumped to “next day” when everything is full

    for day in DAYS:
        # 1) Start with carry-over employees from previous day (if any)
        pool = carry_over + employees
        carry_over = []
        random.shuffle(pool)

        # 2) First pass: try to honor ranked preferences with capacity limits
        for emp in pool:
            if not can_work(emp, day, stats):
                continue
            placed = False
            for pref in EMPLOYEE_PREFERENCES[emp]:
                if has_capacity(day, pref, schedule):
                    assign(emp, day, pref, schedule, stats)
                    placed = True
                    break
            if not placed:
                # Try any other shift with capacity (same day)
                for alt in SHIFTS:
                    if has_capacity(day, alt, schedule):
                        assign(emp, day, alt, schedule, stats)
                        placed = True
                        break
            if not placed:
                # Everything full today → try next day (conflict resolution)
                carry_over.append(emp)

        # 3) Enforce minimum staffing with FAIRNESS (fewest days worked first)
        for shift in SHIFTS:
            needed = max(0, SHIFT_MIN_STAFF - len(schedule[day][shift]))
            if needed == 0:
                continue
            eligible = [
                e for e in employees
                if can_work(e, day, stats) and (day not in stats[e]["assigned_days"])
            ]
            # choose those with the fewest days to spread work across the week
            eligible.sort(key=lambda e: stats[e]["days"])
            for _ in range(needed):
                if not eligible:
                    break
                pick = eligible.pop(0)
                assign(pick, day, shift, schedule, stats)
                # keep list fair as days increase
                eligible.sort(key=lambda e: stats[e]["days"])

        # 4) Optional fair fill to use leftover capacity (still fewest-days first)
        unassigned = [e for e in employees if can_work(e, day, stats)]
        unassigned.sort(key=lambda e: stats[e]["days"])
        for emp in unassigned:
            placed = False
            for pref in EMPLOYEE_PREFERENCES[emp]:
                if has_capacity(day, pref, schedule):
                    assign(emp, day, pref, schedule, stats)
                    placed = True
                    break
            if not placed and not any_capacity(day, schedule):
                carry_over.append(emp)

    return schedule, stats

# ---------- Pretty printing ----------
def print_schedule(schedule, stats):
    print("\n=== FINAL WEEKLY SCHEDULE ===")
    header = f"{'Day':<5} | " + " | ".join(f"{s:<25}" for s in SHIFTS)
    print(header)
    print("-" * len(header))
    for day in DAYS:
        row = [", ".join(schedule[day][s]) if schedule[day][s] else "-" for s in SHIFTS]
        print(f"{day:<5} | " + " | ".join(f"{cell:<25}" for cell in row))

    print("\n=== EMPLOYEE SUMMARY (max 5 days) ===")
    for emp in sorted(stats.keys()):
        days_str = ", ".join(sorted(stats[emp]['assigned_days'], key=lambda d: DAYS.index(d)))
        print(f"{emp:<8}: {stats[emp]['days']} day(s) → {days_str}")

    # Sanity checks
    print("\n=== CONSTRAINT CHECKS ===")
    # 1) One shift per day per employee
    ok_one_shift_per_day = True
    for day in DAYS:
        seen = set()
        for shift in SHIFTS:
            for emp in schedule[day][shift]:
                if emp in seen:
                    ok_one_shift_per_day = False
                seen.add(emp)
    print("One shift per employee per day:", "OK" if ok_one_shift_per_day else "VIOLATION")

    # 2) Weekly cap
    ok_week_cap = all(v["days"] <= MAX_DAYS_PER_EMP for v in stats.values())
    print("Max 5 days per employee:", "OK" if ok_week_cap else "VIOLATION")

    # 3) Minimum staffing
    min_ok = True
    for day in DAYS:
        for shift in SHIFTS:
            if len(schedule[day][shift]) < SHIFT_MIN_STAFF:
                min_ok = False
    print("Minimum 2 employees per shift:", "OK" if min_ok else "COULD NOT FILL (insufficient eligible staff)")

# ---------- Run ----------
if __name__ == "__main__":
    schedule, stats = build_schedule()
    print_schedule(schedule, stats)
